src/librustc_mir/transform/util.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use repr::*;
+
+/// Update basic block ids in all terminators using the given replacements,
+/// useful e.g. after removal of several basic blocks to update all terminators
+/// in a single pass
+pub fn update_basic_block_ids(mir: &mut Mir, replacements: &[usize]) {
+    for bb in mir.all_basic_blocks() {
+        for target in mir.basic_block_data_mut(bb).terminator.successors_mut() {
+            *target = BasicBlock::new(replacements[target.index()]);
+        }
+    }
+}
+
+/// Mass removal of basic blocks to keep the ID-remapping cheap.
+pub fn remove_basic_blocks(mir: &mut Mir, keep: &[bool]) {
+    // Check that we have a usage flag for every block
+    assert_eq!(mir.basic_blocks.len(), keep.len());
+
+    // `mapping` keeps track which new block id refers to which old block id
+    let mut mapping: Vec<_> = (0..mir.basic_blocks.len()).collect();
+    // `replacements` maps the old block ids to the new ones
+    let mut replacements: Vec<_> = mapping.clone();
+
+    for (i, keep) in keep.iter().enumerate().rev() {
+        if !keep {
+            mapping.swap_remove(i);
+            mir.basic_blocks.swap_remove(i);
```

Comment:
I tried to keep the number of swaps proportional to the number of removed blocks, but that isn't worth that much, because I rewrite all targets anyway, and creating the reverse mapping likely eats up any time that has been saved by the reduced number of swaps. So I'll change this.